libglusterfs/src/gf-io-common.c

```diff
@@ -507,6 +507,7 @@ gf_io_thread_pool_start(gf_io_thread_pool_t *pool,
 
     res = gf_io_thread_create(cfg, ids, &created, gf_io_thread_main, &sync);
     if (caa_unlikely(res < 0)) {
+        pending = 0;
         goto done_sync;
     }
     pending = cfg->num_threads - created + 1;
```

Comment:
You cannot set pending = 0 in case of error.

When `gf_io_sync_start()` is called, it's passed the number of entities that will need to be synced. In this case it's ''cfg->num_threads + 1`, representing each one of the newly created threads plus the current thread.

In the ideal case each thread will "signal" 1 of the entities, remaining only the entity corresponding to this thread. But if a thread fails to start, no one will "signal" the entity for that thread. In this case the current thread must also "signal" it, besides its own entity (note that the synchronization object doesn't know what it's synchronizing or whether it has failed or not. It only tracks the number of pending entities to synchronize).

A synchronization is not complete until all entities have been signaled, meaning that `gf_io_sync_done()` will block until then. When a thread fails to start, we need to also "signal" its entity to avoid waiting for something that won't happen.

I think this is the right  way to do it:

```suggestion
    res = gf_io_thread_create(cfg, ids, &created, gf_io_thread_main, &sync);
    pending = cfg->num_threads - created + 1;
    if (caa_unlikely(res < 0)) {
        goto done_sync;
    }
```

Reply:
Thanks for the detailed explanation @xhernandez. Updated the patch as per the suggested change.